### PR TITLE
Expose more shortcut gate types

### PR DESCRIPTION
**Pull request by sjdilkes**

### CODE REVIEW

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 1]
```diff
@@ -751,6 +751,66 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
           "\n\n:return: the new :py:class:`Circuit`",
           py::arg("angle"), py::arg("qubit"))
+      .def(
+          "U1",
+          [](Circuit *circ, const Expr &angle, unsigned qb,
+             const py::kwargs &kwargs) {
+            return add_gate_method_oneparam<unsigned>(
+                circ, OpType::U1, angle, {qb}, kwargs);
+          },
+          "Appends a U1 gate with possible symbolic angle "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:761]:
```suggestion
          "Appends a U1 gate with a possibly symbolic angle "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 2]
```diff
@@ -751,6 +751,66 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          "\n\n:return: the new :py:class:`Circuit`",
+          py::arg("angle"), py::arg("qubit"))
+      .def(
+          "U2",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const unsigned &qb, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<unsigned>(
+                circ, OpType::U2, {angle0, angle1}, {qb}, kwargs);
+          },
+          "Appends a U2 gate with possible symbolic angles "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:772]:
```suggestion
          "Appends a U2 gate with possibly symbolic angles "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 3]
```diff
@@ -751,6 +751,66 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          "\n\n:return: the new :py:class:`Circuit`",
+          py::arg("angle0"), py::arg("angle1"), py::arg("qubit"))
+      .def(
+          "U3",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Expr &angle2, const unsigned &qb, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<unsigned>(
+                circ, OpType::U3, {angle0, angle1, angle2}, {qb}, kwargs);
+          },
+          "Appends a U3 gate with possible symbolic angles "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:783]:
```suggestion
          "Appends a U3 gate with possibly symbolic angles "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 4]
```diff
@@ -751,6 +751,66 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          py::arg("angle0"), py::arg("angle2"), py::arg("angle2"),
+          py::arg("qubit"))
+      .def(
+          "TK1",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Expr &angle2, const unsigned &qb, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<unsigned>(
+                circ, OpType::TK1, {angle0, angle1, angle2}, {qb}, kwargs);
+          },
+          "Appends a TK1 gate with possible symbolic angles "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:795]:
```suggestion
          "Appends a TK1 gate with possibly symbolic angles "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 5]
```diff
@@ -751,6 +751,66 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          "U3",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Expr &angle2, const unsigned &qb, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<unsigned>(
+                circ, OpType::U3, {angle0, angle1, angle2}, {qb}, kwargs);
+          },
+          "Appends a U3 gate with possible symbolic angles "
+          "(specified in half-turns)."
+          "\n\n:return: the new :py:class:`Circuit`",
+          py::arg("angle0"), py::arg("angle2"), py::arg("angle2"),
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:786]:
```suggestion
          py::arg("angle0"), py::arg("angle1"), py::arg("angle2"),
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 6]
```diff
@@ -751,6 +751,66 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          "TK1",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Expr &angle2, const unsigned &qb, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<unsigned>(
+                circ, OpType::TK1, {angle0, angle1, angle2}, {qb}, kwargs);
+          },
+          "Appends a TK1 gate with possible symbolic angles "
+          "(specified in half-turns)."
+          "\n\n:return: the new :py:class:`Circuit`",
+          py::arg("angle0"), py::arg("angle2"), py::arg("angle2"),
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:798]:
```suggestion
          py::arg("angle0"), py::arg("angle1"), py::arg("angle2"),
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 7]
```diff
@@ -751,6 +751,66 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+      .def(
+          "TK2",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Expr &angle2, const unsigned &qb0, const unsigned &qb1,
+             const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<unsigned>(
+                circ, OpType::TK2, {angle0, angle1, angle2}, {qb0, qb1},
+                kwargs);
+          },
+          "Appends a TK2 gate with possible symbolic angles "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:809]:
```suggestion
          "Appends a TK2 gate with possibly symbolic angles "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 8]
```diff
@@ -751,6 +751,66 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          "(specified in half-turns)."
+          "\n\n:return: the new :py:class:`Circuit`",
+          py::arg("angle0"), py::arg("angle2"), py::arg("angle2"),
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:812]:
```suggestion
          py::arg("angle0"), py::arg("angle1"), py::arg("angle2"),
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 9]
```diff
@@ -896,14 +983,48 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          py::arg("qubit1"))
+      .def(
+          "Sycamore",
+          [](Circuit *circ, unsigned qubit0, unsigned qubit1,
+             const py::kwargs &kwargs) {
+            return add_gate_method_noparams<unsigned>(
+                circ, OpType::Sycamore, {qubit0, qubit1}, kwargs);
+          },
+          "Appends a Sycamore gate on the wires for the specified control "
+          "and target qubits."
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1017]:
```suggestion
          "Appends a Sycamore gate on the wires for the specified qubits."
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 10]
```diff
@@ -972,6 +1105,40 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          py::arg("qubit0"), py::arg("qubit1"))
+      .def(
+          "PhasedISWAP",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             unsigned qb0, unsigned qb1, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<unsigned>(
+                circ, OpType::PhasedISWAP, {angle0, angle1}, {qb0, qb1},
+                kwargs);
+          },
+          "Appends a PhasedISWAP gate with symbolic angles (specified in "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1137]:
```suggestion
          "Appends a PhasedISWAP gate with possibly symbolic angles (specified in "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 11]
```diff
@@ -1175,6 +1342,40 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
           "\n\n:return: the new :py:class:`Circuit`",
           py::arg("angle"), py::arg("qubit"))
+      .def(
+          "U1",
+          [](Circuit *circ, const Expr &angle, const Qubit &qb,
+             const py::kwargs &kwargs) {
+            return add_gate_method_oneparam<UnitID>(
+                circ, OpType::U1, angle, {qb}, kwargs);
+          },
+          "Appends a U1 gate with possible symbolic angle "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1352]:
```suggestion
          "Appends a U1 gate with a possibly symbolic angle "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 12]
```diff
@@ -1175,6 +1342,40 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          "\n\n:return: the new :py:class:`Circuit`",
+          py::arg("angle"), py::arg("qubit"))
+      .def(
+          "U2",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Qubit &qb, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<UnitID>(
+                circ, OpType::U2, {angle0, angle1}, {qb}, kwargs);
+          },
+          "Appends a U2 gate with possible symbolic angles "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1363]:
```suggestion
          "Appends a U2 gate with possibly symbolic angles "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 13]
```diff
@@ -1175,6 +1342,40 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          "\n\n:return: the new :py:class:`Circuit`",
+          py::arg("angle0"), py::arg("angle1"), py::arg("qubit"))
+      .def(
+          "U3",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Expr &angle2, const Qubit &qb, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<UnitID>(
+                circ, OpType::U3, {angle0, angle1, angle2}, {qb}, kwargs);
+          },
+          "Appends a U3 gate with possible symbolic angles "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1374]:
```suggestion
          "Appends a U3 gate with possibly symbolic angles "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 14]
```diff
@@ -1175,6 +1342,40 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          "(specified in half-turns)."
+          "\n\n:return: the new :py:class:`Circuit`",
+          py::arg("angle0"), py::arg("angle2"), py::arg("angle2"),
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1377]:
```suggestion
          py::arg("angle0"), py::arg("angle1"), py::arg("angle2"),
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 15]
```diff
@@ -1299,6 +1500,34 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
           "\n\n:return: the new :py:class:`Circuit`",
           py::arg("angle"), py::arg("control_qubit"), py::arg("target_qubit"))
+      .def(
+          "CU1",
+          [](Circuit *circ, const Expr &angle, const Qubit &ctrl,
+             const Qubit &trgt, const py::kwargs &kwargs) {
+            return add_gate_method_oneparam<UnitID>(
+                circ, OpType::CU1, angle, {ctrl, trgt}, kwargs);
+          },
+          "Appends a CU1 gate with a symbolic angle (specified in "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1510]:
```suggestion
          "Appends a CU1 gate with a possibly symbolic angle (specified in "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 16]
```diff
@@ -1299,6 +1500,34 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+      .def(
+          "CU3",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Expr &angle2, const Qubit &ctrl, const Qubit &trgt,
+             const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<UnitID>(
+                circ, OpType::CU3, {angle0, angle1, angle2}, {ctrl, trgt},
+                kwargs);
+          },
+          "Appends a CU3 gate with symbolic angles (specified in "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1524]:
```suggestion
          "Appends a CU3 gate with possibly symbolic angles (specified in "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 17]
```diff
@@ -1320,6 +1549,40 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          py::arg("qubit1"))
+      .def(
+          "Sycamore",
+          [](Circuit *circ, const Qubit &qubit0, const Qubit &qubit1,
+             const py::kwargs &kwargs) {
+            return add_gate_method_noparams<UnitID>(
+                circ, OpType::Sycamore, {qubit0, qubit1}, kwargs);
+          },
+          "Appends a Sycamore gate on the wires for the specified control "
+          "and target qubits."
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1583]:
```suggestion
          "Appends a Sycamore gate on the wires for the specified qubits."
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 18]
```diff
@@ -1396,6 +1670,40 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
+          py::arg("qubit0"), py::arg("qubit1"))
+      .def(
+          "PhasedISWAP",
+          [](Circuit *circ, const Expr &angle0, const Expr &angle1,
+             const Qubit &qb0, const Qubit &qb1, const py::kwargs &kwargs) {
+            return add_gate_method_manyparams<UnitID>(
+                circ, OpType::PhasedISWAP, {angle0, angle1}, {qb0, qb1},
+                kwargs);
+          },
+          "Appends a PhasedISWAP gate with symbolic angles (specified in "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1702]:
```suggestion
          "Appends a PhasedISWAP gate with possibly symbolic angles (specified in "
```

**Review by cqc-alec — CHANGES_REQUESTED** [review]
Please also add a line to the changelog.

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 1 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:761]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 2 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:772]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 3 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:783]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 4 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:795]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 5 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:786]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 6 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:798]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 7 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:809]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 8 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:812]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 9 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1017]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 10 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1137]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 11 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1352]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 12 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1363]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 13 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1374]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 14 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1377]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 15 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1510]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 16 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1524]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 17 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1583]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 18 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp:1702]: done, sorry lots of those.

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 19]
```diff
@@ -972,6 +1104,41 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
           "\n\n:return: the new :py:class:`Circuit`",
           py::arg("control"), py::arg("target_0"), py::arg("target_1"))
+      .def(
+          "ISWAP",
+          [](Circuit *circ, const Expr &angle, unsigned qb0, unsigned qb1,
+             const py::kwargs &kwargs) {
+            return add_gate_method_oneparam<unsigned>(
+                circ, OpType::ISWAP, angle, {qb0, qb1}, kwargs);
+          },
+          "Appends an ISWAP gate with a symbolic angle (specified in "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1114]:
```suggestion
          "Appends an ISWAP gate with a possibly symbolic angle (specified in "
```

File: pytket/binders/circuit/Circuit/add_op.cpp [hunk 20]
```diff
@@ -1396,6 +1669,41 @@ void init_circuit_add_op(py::class_<Circuit, std::shared_ptr<Circuit>> &c) {
...
           "\n\n:return: the new :py:class:`Circuit`",
           py::arg("control"), py::arg("target_0"), py::arg("target_1"))
+      .def(
+          "ISWAP",
+          [](Circuit *circ, const Expr &angle, const Qubit &qb0,
+             const Qubit &qb1, const py::kwargs &kwargs) {
+            return add_gate_method_oneparam<UnitID>(
+                circ, OpType::ISWAP, angle, {qb0, qb1}, kwargs);
+          },
+          "Appends an ISWAP gate with a symbolic angle (specified in "
```
Comment (cqc-alec) [inline pytket/binders/circuit/Circuit/add_op.cpp:1679]:
```suggestion
          "Appends an ISWAP gate with a possibly symbolic angle (specified in "
```

**Review by cqc-alec — CHANGES_REQUESTED** [review]
Thanks, don't forget to update the changelog!

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 20 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp]: done

File: pytket/binders/circuit/Circuit/add_op.cpp (hunk 19 above)
Comment (sjdilkes) [reply pytket/binders/circuit/Circuit/add_op.cpp]: done

**Comment by sjdilkes** [comment]
> Thanks, don't forget to update the changelog!

I've updated the changelog now - I've also cheekily updated it for the last PR which I forgot to update it for.

File: pytket/docs/changelog.rst [hunk 21]
```diff
@@ -4,6 +4,13 @@ Changelog
 1.8.1 (November 2022)
 ---------------------
 
+Minor new features:
+
+* Added shortcuts for adding ``U1``, ``U2``, ``U3``, ``TK1``, ``TK2``, ``CU1``, 
+  ``CU3``, ``ISWAP``, ``PhasedISWAP``, ``ESWAP``, ``PhasedX``, ``FSim``, ``Sycamore``
+  and ``ISWAPMax`` gates to a ``pytket`` ``Circuit``.
+* New ``Circuit`` methods ``n_1qb_gates``, ``n_2qb_gates``, ``n_nqb_gates``.
```
Comment (cqc-alec) [inline pytket/docs/changelog.rst:12]: These should be under 1.9.0 (unreleased).

Comment (sjdilkes) [reply pytket/docs/changelog.rst:12]: done

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**